numpy/core/src/multiarray/descriptor.c

```diff
@@ -1072,7 +1072,13 @@ _convert_from_dict(PyObject *obj, int align)
                 goto fail;
             }
             offset = PyInt_AsLong(off);
-            PyTuple_SET_ITEM(tup, 1, off);
+            if (PyErr_Occurred()) {
```

Comment:
Two rather unrelated side points I am not sure about:
 1. What about floats, etc.? Maybe it would make sense to use the Index conversion stuff?
 2. Where do we check `int` size? We only support integer sized offsets (since dtype is limited to int).